docs/adr/0001-move-to-event-sourcing-based-system.md

```diff
@@ -0,0 +1,88 @@
+
+As we have considered this, it is obvious that there is a strong resemblence between our workflow engine and a system based on an event sourcing architecture:
+- Workflow instances are our Aggregate Roots.
+- Triggers are our Commands.
+- Transitions are our Events.
+- Actions are the means by which our Events are published to subscribers.
+
+As a result, we are considering re-working the internals of the workflow engine to more closely follow an event sourcing approach.
+
+## Decision
+
+We will rebuild the workflow engine using an event sourcing approach to model workflow instances.
+
+## Consequences
+
+### Workflow instances are now stored in an event store
+
+There will likely only be two types of event that are generated by a workflow instance - one for when the instance is first created, and then another type representing transitions.
+
+### Workflow transition execution logic will be broken up
+
+At present, the processing sequence for a transition is as follows:
+- Verify exit conditions from current state
+- Verify transition conditions
+- Verify entry conditions on new state
+- Execute exit actions on current state
+- Execute transition actions
+- Execute entry actions on new state
+
+Since the actions represent things that need to happen as a result of the transition, the processing will need to change, and will look more like the following:
+- Verify exit conditions from current state
+- Verify transition conditions
+- Verify entry conditions on new state
+- Create a transition event containing all the information pertaining to the current transition - essentially transition Id and trigger data and store it in the event store (which will result in it being published to all subscribers).
+
+Separately, the workflow engine will subscribe to its own events. On receiving an event the engine will use the supplied data to:
+- Load the workflow instance (so that we can supply context data to actions)
+- Load the workflow definition
+- Execute exit actions on the source state
+- Execute transition actions
+- Execute entry actions on the target state
+
+This will allow us to retain the ability to include actions as part of a workflow definition - useful because one of the big benefits of the workflow-based approach is that it makes it easier to see what happens as a result of transitions.
+
+### Consumers can choose whether to use workflow actions or event store subscriptions when modelling their workflows
+
+
+
+### Workflow actions should no longer be able to modify workflow instance context data
+
+
+
+### We need a new category of "mutation" that happens as part of workflow state changes
+
+
+
+### Replay becomes possible
+
+
+
+### Workflow instances become more resiliant
+
+
+
+### We have an audit trail
+
+
+### Eventual consistency becomes a bigger issue for clients to manage
```

Comment:
Eventual consistency is an issue that clients can no longer ignore and hope for the best.

Or rather, we make it clear that if they continue to do so, they will have a problem.

The issue is "flow" - we do not know the consequences of an action without some kind of closed loop that the client knows about. This could/should be modelled in the workflow. (Transition to "working on this stuff state" wait for all the "stuff" to finish, transition to "wherever you would have gone directly") - the UI can subscribe to those transitions (signalR?) to find out when it should update, and, of course, present suitable "waiting" UI for the transitional state if it comes back during the process.